### PR TITLE
nixos/boot: add postResumeCommands option

### PR DESCRIPTION
**Pull request by accelbread**
###### Description of changes

Adds a postResumeCommands option to the initramfs to allow inserting code to execute after the device has attempted to resume, and before filesystems are mounted.

 This allows to inject code for operations like wiping the rootfs on boot; if those were instead put in postDeviceCommands, on a hibernated device, they would execute before the device resumes from hibernation. The next currently available spot is postMountCommands which is too late.

Currently, impermanence setups recommend using postDeviceCommands, see:
https://grahamc.com/blog/erase-your-darlings/
https://mt-caret.github.io/blog/posts/2020-06-29-optin-state.html

###### Things done

- Built on platform(s)
  - [X] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- [ ] For non-Linux: Is `sandbox = true` set in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
- [X] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [X] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [ ] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [23.11 Release Notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2311.section.md) (or backporting [23.05 Release notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2305.section.md))
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [X] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

### CODE REVIEW

**Review by bjornfor — APPROVED** [review]
LGTM.

**Review by Luflosi — APPROVED** [review]

**Review by ElvishJerricco — CHANGES_REQUESTED** [review]
This should be added to the list of options that are asserted to be unset for systemd-initrd:

https://github.com/NixOS/nixpkgs/blob/4285a2a67daf39e63d9564a47773a1c2081c36a8/nixos/modules/system/boot/systemd/initrd.nix#L347-L366

**Comment by accelbread** [comment]
Added it to the systemd-initrd asserts

**Review by ElvishJerricco — APPROVED** [review]

**Comment by NireBryce** [comment]
This feels badly named, or it's supposed to be in the conditional block above it, since as far as I can tell this will execute regardless of hibernate, but it's described as only triggering after resume, not regardless of resume.  (there's also the possibility that boot counts as resume, but maybe it needs to be more clear?)

**Comment by accelbread** [comment]
This is for injecting code into stage1 after resume handling is complete. This is for code that is post device setup, but should only run after attempting to resume. Naming matches other options.

I use this to delete my root subvolume. Should be done on boot after devices, but not if we are waking from hibernate. Using postDeviceCommands runs the hook, and then after resumes to a system using that deleted subvolume...

**Comment by NireBryce** [comment]
Okay, I get it now.  (I'm also here because of impermanence)

But in terms of clarity, the nixpkgs description seems accidentally misleading:
```nix
description =  ''shell command to be executed immediately after attempting to resume'';
```
 ([nixpkgs/modules/system/boot/stage-1.nix](https://github.com/NixOS/nixpkgs/blob/45cbf8d42fff56fcac72bbd31292eb381134be93/nixos/modules/system/boot/stage-1.nix#L526))

**Comment by accelbread** [comment]
maybe "attempting resume" could be "handling resume"?

**Comment by ElvishJerricco** [comment]
I would reword it as "Shell commands to be executed immediately after attempting to resume and resume did not occur."
